cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,261 @@
+		updateTime:              timeNow(),
+		timeNow:                 timeNow,
+		perOperationRateLimiter: perOperationRateLimiter,
+	}
+}
+
+// Withdraw credits from the creditAccruer for the given operationName with an
+// upper limit of maxWithdrawal.
+func (accruer *creditAccruer) Withdraw(operationName string, maxWithdrawal float64) float64 {
+	var rateLimiter *tokenBucket
+	if accruer.perOperationRateLimiter.Has(operationName) ||
+		accruer.perOperationRateLimiter.IsFull() {
+		rateLimiter = accruer.perOperationRateLimiter.Get(operationName).(*tokenBucket)
+	} else {
+		rateLimiter = newTokenBucket(accruer.options, accruer.timeNow)
+		accruer.perOperationRateLimiter.Set(operationName, rateLimiter)
+	}
+	credits := rateLimiter.Withdraw(maxWithdrawal)
+	return credits
+}
+
+// CreditStore manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type CreditStore interface {
+	io.Closer
+	// Withdraw deducts as many credits from the service on behalf of the client
+	// without exceeding MaxClientBalance. The client is identified by a unique
+	// clientID string. The credits are deducted from the specific operation on
+	// the service if present, otherwise from the default credit pool.
+	Withdraw(clientID string, serviceName string, operationName string) float64
+	// Spend credits already allocated to this client on a given operation.
+	Spend(clientID string, operationName string, credits float64) bool
+}
+
+type creditStore struct {
+	sync.Mutex
+	creditAccruers map[string]*creditAccruer
+	clients        map[string]*client
+	options        CreditAccruerOptions
+	ttl            time.Duration
+	timeNow        func() time.Time // For testing
+	ticker         *time.Ticker
+	done           chan bool
+	waitGroup      sync.WaitGroup
+}
+
+// NewCreditStore creates a new creditStore and returns it to the caller.
+// options should be the default values passed to new creditAccruers.
+// ttl should be the TTL for all creditAccruers and clients.
+func NewCreditStore(options CreditAccruerOptions, ttl time.Duration) CreditStore {
+	s := &creditStore{
+		creditAccruers: map[string]*creditAccruer{},
+		clients:        map[string]*client{},
+		options:        options,
+		ttl:            ttl,
+	}
+	s.timeNow = time.Now
+	s.ticker = time.NewTicker(10 * time.Second)
+	s.done = make(chan bool)
+	s.waitGroup.Add(1)
+	go func() {
+		defer s.waitGroup.Done()
+		for {
+			select {
+			case <-s.ticker.C:
+				s.purgeExpired()
+			case <-s.done:
+				return
+			}
+		}
+	}()
+	return s
+}
+
+func (s *creditStore) Withdraw(clientID string, serviceName string, operationName string) float64 {
+	s.Lock()
+	defer s.Unlock()
+	c := s.findOrCreateClient(clientID)
+	accruer := s.findOrCreateCreditAccruer(serviceName)
```

Comment:
nit: "credit accruer" doesn't sound like a good term in this file, since it does not explain how, for example, it's different from credit store. It would be nice to minimize the number of terms. Isn't it just a collection of token buckets per operation?

Reply:
True. I was considering using a financial analogy throughout to simplify things. The earlier design used the term credit accruer, so I kept it, but I'm thinking this might work better:

* `CreditBank` to replace `CreditStore`
* `account` to replace `creditAccruer`

Then at least it is easy to explain that several clients can share an account in a bank. Let me know if you prefer using existing terms or the analogy I just described (or another analogy).

Reply:
How about Throttler as the main class? Throttling is its primary business function for us, and the "credits" are an implementation detail.

I like `account`.

Reply:
Cool will do that